runtime/observed.go

```diff
@@ -0,0 +1,65 @@
+package runtime
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	xfnv1alpha1 "github.com/crossplane/crossplane/apis/apiextensions/fn/io/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ObservedResources struct {
+	resources []Resource
+	composite xfnv1alpha1.ObservedComposite
+}
+
+// GetFromKubeObject gets the k8s resource o from a provider kubernetes object kon (Kube Object Name)
+// from the observed array of the FunctionIO.
+func (o *ObservedResources) GetFromKubeObject(ctx context.Context, obj client.Object, kon string) error {
+	ko, err := getKubeObjectFrom(ctx, &o.resources, kon)
+	if err != nil {
+		return err
+	}
+	return fromKubeObject(ctx, ko, obj)
+}
+
+// GetManagedResource unmarshalls the managed resource with the given name into the given object.
+// It reads from the Observed array.
+func (o *ObservedResources) GetManagedResource(ctx context.Context, resName string, obj client.Object) error {
```

Comment:
Nitpick: The signature is also not consistent (ctx, obj, name vs ctx, name, obj)